### PR TITLE
replace *zio.Writer with zbuf.WriteCloser

### LINKED ISSUE #1235: rework zio design pattern so it's easier to add zst

### PR DESCRIPTION
**Pull request by mccanne**
This commit cleans up the zio design pattern to allow generic
writers to be closed instead of doing a flush followed by reaching
around the writer to close the underlyinig object.

This will make integrating the zst writer easier because it needs
to be closed to finish up writing metadata and a trailer.  I was
fighting with the existing pattern to wire it all up and this change
will make it all straightforward.

The previous flush-based approach had merits but this seems
easier to understand and work with, albeit we had to add Close
stubs to all of the zio writers.

We also mooved zbuf/zng.go to zbuf/zbuf.go as the previous
name was historical and no longer accurate, and deleted
zbuf.WriteFlusher as it is no longer used.

Finally, this refactoring exposed a problem in the fs.ReplaceFile
call in storage filestore, which was doing a storage copy without
any write buffering, so this is fixed.

Fixed #1235

### CODE REVIEW

File: zio/tableio/writer.go
Comment (alfred-landrum) [inline zio/tableio/writer.go:95]:
```suggestion
	err := w.flush()
	if closeErr := w.writer.Close(); closeErr != nil && err == nil {
		err = closeErr
	}
	return err
```

Comment (mccanne) [reply zio/tableio/writer.go:95]: This the same but with different names.  I was using Noah's pattern "firstErr" but I have a slight preference for this since the idiomatic "err" stays in the outer scope.

File: zio/zngio/writer.go
Comment (mccanne) [reply zio/zngio/writer.go:45]: Yes, it was supposed to be 
```
firstErr = err
```
instead of return, but II will change to your pattern (see comment below)

File: zio/tableio/writer.go
Comment (mccanne) [reply zio/tableio/writer.go:95]: btw, microindex.Writer uses the firstErr pattern so we should probably change it to be consistent.  I'll put up another PR.

Comment (nwt) [reply zio/tableio/writer.go]: Not that you asked, but I like
```go
	err := w.flush()
	if err2 := w.writer.Close(); err == nil {
		err = err2
	}
	return err

```
for this situation.

Comment (mccanne) [reply zio/tableio/writer.go]: That's where we landed but do you want me to change the var name `closeErr` to `err2`?

Comment (mccanne) [reply zio/tableio/writer.go]: https://github.com/brimsec/zq/pull/1236/commits/b803091c1ae2f6295d24547cec707eb33804e901

Comment (nwt) [reply zio/tableio/writer.go]: > That's where we landed but do you want me to change the var name closeErr to err2?

Go with your preference.

File: zqd/storage/filestore/filestore.go
Comment (alfred-landrum) [inline zqd/storage/filestore/filestore.go:123]: 👍

**Review by alfred-landrum — APPROVED** [review]
LGTM, would wait to see if @nwt has further comments

File: zio/zio.go
Comment (nwt) [inline zio/zio.go:70]: Nit: I think this would be a little nicer if exposed as a function rather than a type, like io/ioutil.NopCloser, but we can do that later.
```suggestion
// NoCloser returns an io.WriteCloser with a no-op Close method wrapping w.
func NoCloser(w io.Writer) io.WriteCloser { return &noCloser{w} }

type noCloser struct{ io.Writer }

func (_ *noCloser) Close() error { return nil }
```

**Review by nwt — APPROVED** [review]

File: zio/zio.go
Comment (mccanne) [reply zio/zio.go:70]: Thanks, I'll make the change before I merge and follow the go pattern by changing the to NopCloser.

**Pull request merged**
